--- /mnt/share/linux/scripts/Makefile.build	2009-06-09 11:26:21.000000000 +0900
+++ /mnt/share/linux_orig/scripts/Makefile.build	2009-06-05 13:11:23.000000000 +0900
@@ -178,7 +178,7 @@
 
 cmd_cc_o_c = $(CC) $(c_flags) -c -o $(@D)/.tmp_$(@F) $<
 cmd_modversions =							\
-	if false && $(OBJDUMP) -h $(@D)/.tmp_$(@F) | grep -q __ksymtab; then	\
+	if $(OBJDUMP) -h $(@D)/.tmp_$(@F) | grep -q __ksymtab; then	\
 		$(CPP) -D__GENKSYMS__ $(c_flags) $<			\
 		| $(GENKSYMS) $(if $(KBUILD_SYMTYPES),			\
 			      -T $(@D)/$(@F:.o=.symtypes)) -a $(ARCH)	\
